lib/northbound.c

```diff
@@ -497,52 +496,154 @@ static void nb_config_diff_deleted(const struct lyd_node *dnode, uint32_t *seq,
 		 * do is to call the "destroy" callbacks of their child nodes
 		 * when applicable (i.e. optional nodes).
 		 */
-		LY_TREE_FOR (dnode->child, child) {
+		LY_LIST_FOR (lyd_child(dnode), child) {
 			nb_config_diff_deleted(child, seq, changes);
 		}
 	}
 }
 
+static int nb_lyd_diff_get_op(const struct lyd_node *dnode)
+{
+	const struct lyd_meta *meta;
+	LY_LIST_FOR (dnode->meta, meta) {
+		if (strcmp(meta->name, "operation")
+		    || strcmp(meta->annotation->module->name, "yang"))
+			continue;
+		return lyd_get_meta_value(meta)[0];
+	}
+	return 'n';
+}
+
+static inline void nb_config_diff_dnode_log_path(const char *context,
+						 const char *path,
+						 const struct lyd_node *dnode)
+{
+	if (dnode->schema->nodetype & LYD_NODE_TERM)
+		zlog_debug("nb_config_diff: %s: %s: %s", context, path,
+			   lyd_get_value(dnode));
+	else
+		zlog_debug("nb_config_diff: %s: %s", context, path);
+}
+
+static inline void nb_config_diff_dnode_log(const char *context,
+					    const struct lyd_node *dnode)
+{
+	if (!dnode) {
+		zlog_debug("nb_config_diff: %s: NULL", context);
+		return;
+	}
+
+	char *path = lyd_path(dnode, LYD_PATH_STD, NULL, 0);
+	nb_config_diff_dnode_log_path(context, path, dnode);
+	free(path);
+}
+
 /* Calculate the delta between two different configurations. */
 static void nb_config_diff(const struct nb_config *config1,
```

Comment:
So, I've been over this function a few times. To make sure it's correct. Before I get to that, it's worth noting that we pass all topotests, changes aren't missed etc, so I've got some backing for my belief that things are correct. :)

Now to your question about the no-op specified inheritance. The thing you have to notice here is that we do not descend into the diff tree beyond the first node that has an operation defined. That is what `LYD_TREE_DFS_continue = 1; /* done with sub-tree*/` is doing. There's no way to actually reach a node that inherits a parent's operation b/c we terminate the decent at the first node that has the create or a delete operation. Change operations only occur on leaf nodes so again nothing to inherit below. 

The other reason I've looked multiple times at this function primarily has to do with wanting to try to optimize it in the direction that you mention; however, I do not believe that this should be done with this merge. This can be looked at being done separately later on. The current code basically mimics the previous libyang1 code and so matches the current overall northbound design-- a good first step for the transition.